app/controllers/SelectNINOLetterAddressController.scala

```diff
@@ -83,96 +84,68 @@ class SelectNINOLetterAddressController @Inject()(
-                          identifierType = pdv.map(_.CRN.getOrElse("")).getOrElse(""),
-                          pageErrorGeneratedFrom = Some("/postcode"),
-                          errorStatus = Some(responseStatus.toString),
-                          errorReason = Some(responseMessage)
-                        ))
-                      case Failure(ex) => logger.warn(ex.getMessage)
-                    }
-                    Redirect(routes.SendLetterErrorController.onPageLoad(mode))
-                  case TechnicalIssueResponse(responseStatus, responseMessage) =>
-                    personalDetailsValidationService.getPersonalDetailsValidationByNino(nino).onComplete {
-                      case Success(pdv) =>
-                        auditService.audit(AuditUtils.buildAuditEvent(pdv.flatMap(_.personalDetails),
-                          auditType = "FindYourNinoError",
-                          validationOutcome = pdv.map(_.validationStatus).getOrElse(""),
-                          identifierType = pdv.map(_.CRN.getOrElse("")).getOrElse(""),
-                          pageErrorGeneratedFrom = Some("/postcode"),
-                          errorStatus = Some(responseStatus.toString),
-                          errorReason = Some(responseMessage)
-                        ))
-                      case Failure(ex) => logger.warn(ex.getMessage)
-                    }
-                    Redirect(routes.TechnicalErrorController.onPageLoad())
-                  case _ =>
-                    logger.warn("Unknown NPS FMN API response")
-                    Redirect(routes.TechnicalErrorController.onPageLoad())
+      personalDetailsValidationService.getPersonalDetailsValidationByNino(nino).flatMap(pdvData =>
+        form.bindFromRequest().fold(
+          formWithErrors =>
+            Future.successful(BadRequest(view(formWithErrors, mode, getPostCode(pdvData)))),
+          value => {
+            request.userAnswers.set(SelectNINOLetterAddressPage, value) match {
+              case Failure(_) => Future.successful(Redirect(routes.TechnicalErrorController.onPageLoad()))
+              case Success(uA) =>
+                sessionRepository.set(uA)
+                uA.get(SelectNINOLetterAddressPage) match {
+                  case Some(SelectNINOLetterAddress.NotThisAddress) =>
+                    Future.successful(Redirect(navigator.nextPage(SelectNINOLetterAddressPage, mode, uA)))
+                  case Some(SelectNINOLetterAddress.Postcode) =>
+                    sendLetter(nino, pdvData, value.toString, uA, mode)
                 }
             }
           }
-        }.flatten
+        )
       )
   }
 
+  private def sendLetter(nino: String, pdvData: Option[PDVResponseData], value: String, uA: UserAnswers, mode: Mode)
+                        (implicit headerCarrier: HeaderCarrier): Future[Result] = {
+    npsFMNService.sendLetter(nino, getNPSFMNRequest(pdvData)) map {
+      case LetterIssuedResponse() =>
+        for {
```

Comment:
We do not need the for comprehension in this block. The block could be rewritten using a `map` as follows. 

`getIndividualDetailsAddress(IndividualDetailsNino(nino)) map {
          case Right(idAddress) => auditHelper.findYourNinoOnlineLetterOption(pdvData, idAddress, value)
          case _                            => throw new IllegalArgumentException("Could not get individuals address")
}
Redirect(navigator.nextPage(SelectNINOLetterAddressPage, mode, uA))`

Also do we want to handle the case we can't pull the address? At the moment we are swallowing this error (left) and only auditing if we can get the address? Is this intended? Should we always have an address?

Reply:
@rishisachania  @Kitkatann  - Are we currently deliberately ignoring the LEFT (error) and not auditing where we have no address? Feels wrong?